mantidproject/mantid
# Muon Analysis differences

**Description of work.**
This adds a difference table to the muon analysis and frequency domain analysis GUI's. The difference table allows users to look at either `group1 - group2` or `pair1 - pair2` (pairs are ratios of groups).

The tests will be added in https://github.com/mantidproject/mantid/issues/30558 and some code clean up will be done too.



**To test:**


Open muon analysis
Set to HIFI
load run `84447` (has 4 periods)
go to the grouping tab
at the bottom of the tab will be a new table
by default it will have two values `pair1_minus_2` and `pair_3_minus_4`
ticking the analysis checkbox will result in the difference being added to the plot
If you go to fitting or sequential fitting, the selected difference will be available 

Check you can create a new pair difference (add button)
Change the difference (e.g. long2 to long3) - if the plot is being shown, it will update
Check you can delete a difference (minus button)

Above the difference table are a pair of radio buttons - change it to group
Check you can add, remove, select and edit group differences

Check that if you create a new pair or group that the new one is available in the appropriate difference table (and not the other)

Check clicking default at the top of the grouping tab, resets correctly
Check clicking clear, clears all of 4 tables
Check you can save a custom set of differences and that you can load them back in (press clear in between)

Click default on the grouping tab
In the run box enter `84447, 85557` (the second run is single period)
A lot of yellow highlights will appear (all of the default differences will be yellow)
Click the group radio button
Add a group difference 
If you do (fwd1, fwd2) it will be yellow, if you do `fwd1`, `bwd1` it will be white.

Close muon analysis
Open frequency domain analysis
Check the above still works (it should, with the exception of fitting and sequential fitting)

When you tick the analysis checkbox for the differences it will now be available as a workspace in the transform tab (FFT)

Fixes #30420. 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
